tests/e2e/src/api/midnight.rs

```diff
@@ -46,25 +53,91 @@ pub async fn subscribe_to_cngd_registration_extrinsic(
 								for utxo in utxos {
 									let utxo_tx_id = utxo.header.tx_hash.0;
 									if utxo_tx_id == tx_id {
-										println!("*** Found UTXO with matching registration tx hash: 0x{} ***", hex::encode(tx_id));
+										println!(
+											"*** Found UTXO with matching tx hash: 0x{} ***",
+											hex::encode(tx_id)
+										);
 										return Ok(events);
 									} else {
-										println!("Tx hash 0x{} does not match expected registration tx hash 0x{}", hex::encode(utxo_tx_id), hex::encode(tx_id));
+										println!(
+											"Tx hash 0x{} does not match expected tx hash 0x{}",
+											hex::encode(utxo_tx_id),
+											hex::encode(tx_id)
+										);
 									}
 								}
 							}
 						}
+					},
 					_ => {
 						continue;
 					},
 				}
 			}
 		}
-		Err("Did not find registration event".into())
-	}).await;
+		Err("Did not find native token observation event".into())
+	})
+	.await;
 
 	match result {
 		Ok(res) => res,
-		Err(_) => Err("Timeout waiting for registration event".into()),
+		Err(_) => Err("Timeout waiting for native token observation event".into()),
+	}
+}
+
+pub fn calculate_nonce(prefix: &[u8], tx_hash: [u8; 32], tx_index: u16) -> String {
+	let tx_index_bytes = tx_index.to_be_bytes();
+	let mut data = Vec::new();
+	data.extend_from_slice(prefix);
+	data.extend_from_slice(&tx_hash);
+	data.extend_from_slice(&tx_index_bytes);
+
+	let mut hasher = Blake2bVar::new(32).unwrap();
+	hasher.update(&data);
+	let nonce = hasher.finalize_boxed();
+	hex::encode(&nonce)
+}
```

Comment:
We have this code already somewhere in node, can we just make that function public or move it to a util lib?

Reply:
The hash is inside the pallet logic, and isn't exposed as a function: https://github.com/midnightntwrk/midnight-node/blob/dae4892d3fd9adc32feb08e34a933a949a81af8e/pallets/native-token-observation/src/lib.rs#L432-L435

We can move it out into an external function - probably the best place would be in the cnight primitives crate

We can do it in a separate ticket